src/librustc_mir/transform/add_retag.rs

```diff
@@ -0,0 +1,169 @@
+        ty::Adt(adt, substs) =>
+            adt.variants.iter().any(|v| v.fields.iter().any(|f|
+                has_reference(f.ty(tcx, substs), tcx)
+            )),
+        // Conservative fallback
+        _ => true,
+    }
+}
+
+impl MirPass for AddRetag {
+    fn run_pass<'a, 'tcx>(&self,
+                          tcx: TyCtxt<'a, 'tcx, 'tcx>,
+                          _src: MirSource,
+                          mir: &mut Mir<'tcx>)
+    {
+        if !tcx.sess.opts.debugging_opts.mir_emit_retag {
+            return;
+        }
+        let (span, arg_count) = (mir.span, mir.arg_count);
+        let (basic_blocks, local_decls) = mir.basic_blocks_and_local_decls_mut();
+        let needs_retag = |place: &Place<'tcx>| {
+            is_local(place) && has_reference(place.ty(&*local_decls, tcx).to_ty(tcx), tcx)
+        };
+
+        // PART 1
+        // Retag arguments at the beginning of the start block.
+        {
+            let source_info = SourceInfo {
+                scope: OUTERMOST_SOURCE_SCOPE,
+                span: span, // FIXME: Consider using just the span covering the function
+                            // argument declaration.
+            };
+            // Gather all arguments, skip return value.
+            let places = local_decls.iter_enumerated().skip(1).take(arg_count)
+                    .map(|(local, _)| Place::Local(local))
+                    .filter(needs_retag)
+                    .collect::<Vec<_>>();
+            // Emit their retags.
+            basic_blocks[START_BLOCK].statements.splice(0..0,
+                places.into_iter().map(|place| Statement {
+                    source_info,
+                    kind: StatementKind::Retag { fn_entry: true, place },
+                })
+            );
+        }
+
+        // PART 2
+        // Retag return values of functions.
+        // We collect the return destinations because we cannot mutate while iterating.
+        let mut returns: Vec<(SourceInfo, Place<'tcx>, BasicBlock)> = Vec::new();
+        for block_data in basic_blocks.iter_mut() {
+            match block_data.terminator {
+                Some(Terminator { kind: TerminatorKind::Call { ref destination, .. },
+                                  source_info }) => {
+                    // Remember the return destination for later
+                    if let &Some(ref destination) = destination {
+                        if needs_retag(&destination.0) {
+                            returns.push((source_info, destination.0.clone(), destination.1));
+                        }
+                    }
+                }
+                _ => {
+                    // Not a block ending in a Call -> ignore.
+                    // `Drop` is also a call, but it doesn't return anything so we are good.
+                }
+            }
+        }
+        // Now we go over the returns we collected to retag the return values.
+        for (source_info, dest_place, dest_block) in returns {
```

Comment:
no splice here?

Reply:
It's just a single statement being inserted, how would `splice` help?

Reply:
ups, I misread that outer loop